File: src/limitedwip/common/settings/SettingsForm.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package limitedwip.common.settings;

import com.intellij.ide.BrowserUtil;
import com.intellij.ui.components.labels.LinkLabel;
import com.intellij.ui.components.labels.LinkListener;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

public class SettingsForm {
	public JPanel root;

	private JCheckBox watchdogEnabled;
	private JComboBox maxLinesInChange;
	private JComboBox notificationInterval;
	private JCheckBox showRemainingInToolbar;
	private JCheckBox disableCommitsAboveThreshold;

	private JCheckBox autoRevertEnabled;
	private JComboBox minutesTillRevert;
	private JCheckBox notifyOnRevert;
	private JCheckBox showTimerInToolbar;
	private LinkLabel<Void> openReadme;

	private final LimitedWIPSettings initialState;
	private LimitedWIPSettings currentState;
	private boolean isUpdatingUI;

	public SettingsForm(LimitedWIPSettings initialState) {
		this.initialState = initialState;
		this.currentState = new LimitedWIPSettings();
		currentState.loadState(initialState);
		updateUIFromState();

		ActionListener commonActionListener = new ActionListener() {
			@Override public void actionPerformed(@NotNull ActionEvent event) {
				updateStateFromUI();
				updateUIFromState();
			}
		};

		watchdogEnabled.addActionListener(commonActionListener);
		maxLinesInChange.addActionListener(commonActionListener);
		notificationInterval.addActionListener(commonActionListener);
		showRemainingInToolbar.addActionListener(commonActionListener);
		disableCommitsAboveThreshold.addActionListener(commonActionListener);

		autoRevertEnabled.addActionListener(commonActionListener);
		minutesTillRevert.addActionListener(commonActionListener);
		notifyOnRevert.addActionListener(commonActionListener);
		showTimerInToolbar.addActionListener(commonActionListener);

		openReadme.setListener(new LinkListener<Void>() {
			@Override public void linkSelected(LinkLabel aSource, Void aLinkData) {
				BrowserUtil.open("https://github.com/dkandalov/limited-wip/blob/master/README.md");
			}
		}, null);
	}

	public void updateUIFromState() {
		if (isUpdatingUI) return;
		isUpdatingUI = true;

		watchdogEnabled.setSelected(currentState.watchdogEnabled);
		maxLinesInChange.setSelectedItem(String.valueOf(currentState.maxLinesInChange));
		notificationInterval.setSelectedItem(String.valueOf(currentState.notificationIntervalInMinutes));
		showRemainingInToolbar.setSelected(currentState.showRemainingChangesInToolbar);
		disableCommitsAboveThreshold.setSelected(currentState.disableCommitsAboveThreshold);

		autoRevertEnabled.setSelected(currentState.autoRevertEnabled);
		minutesTillRevert.setSelectedItem(String.valueOf(currentState.minutesTillRevert));
		notifyOnRevert.setSelected(currentState.notifyOnRevert);
		showTimerInToolbar.setSelected(currentState.showTimerInToolbar);

		minutesTillRevert.setEnabled(currentState.autoRevertEnabled);
		notifyOnRevert.setEnabled(currentState.autoRevertEnabled);
		showTimerInToolbar.setEnabled(currentState.autoRevertEnabled);
		maxLinesInChange.setEnabled(currentState.watchdogEnabled);
		notificationInterval.setEnabled(currentState.watchdogEnabled);
		showRemainingInToolbar.setEnabled(currentState.watchdogEnabled);
		disableCommitsAboveThreshold.setEnabled(currentState.watchdogEnabled);

		isUpdatingUI = false;
	}

	private void updateStateFromUI() {
		try {
			currentState.watchdogEnabled = watchdogEnabled.isSelected();
			Integer lineCount = Integer.valueOf((String) maxLinesInChange.getSelectedItem());
			if (LimitedWIPSettings.changedLinesRange.isWithin(lineCount)) {
				currentState.maxLinesInChange = lineCount;
			}
			Integer minutes = Integer.valueOf((String) notificationInterval.getSelectedItem());
			if (LimitedWIPSettings.notificationIntervalRange.isWithin(minutes)) {
				currentState.notificationIntervalInMinutes = minutes;
			}
			currentState.showRemainingChangesInToolbar = showRemainingInToolbar.isSelected();
			currentState.disableCommitsAboveThreshold = disableCommitsAboveThreshold.isSelected();

			currentState.autoRevertEnabled = autoRevertEnabled.isSelected();
			minutes = Integer.valueOf((String) minutesTillRevert.getSelectedItem());
			if (LimitedWIPSettings.minutesToRevertRange.isWithin(minutes)) {
				currentState.minutesTillRevert = minutes;
			}
			currentState.notifyOnRevert = notifyOnRevert.isSelected();
			currentState.showTimerInToolbar = showTimerInToolbar.isSelected();

		} catch (NumberFormatException ignored) {
		}
	}

	public boolean isModified() {
		return !currentState.equals(initialState);
	}

	public LimitedWIPSettings applyChanges() {
		initialState.loadState(currentState);
		return initialState;
	}

	public void resetChanges() {
		currentState.loadState(initialState);
	}
}

File: src/limitedwip/common/settings/SettingsForm.kt
package limitedwip.common.settings

import com.google.common.collect.HashBiMap
import com.intellij.ide.BrowserUtil
import com.intellij.ui.IdeBorderFactory
import com.intellij.ui.RawCommandLineEditor
import com.intellij.ui.components.labels.LinkLabel
import com.intellij.util.execution.ParametersListUtil
import limitedwip.common.settings.LimitedWipSettings.Companion.isValidChangedSizeRange
import limitedwip.common.settings.LimitedWipSettings.Companion.isValidMinutesTillRevert
import limitedwip.common.settings.LimitedWipSettings.Companion.isValidNotificationInterval
import limitedwip.common.settings.TcrAction.*
import java.awt.event.ActionEvent
import javax.swing.JCheckBox
import javax.swing.JComboBox
import javax.swing.JPanel
import javax.swing.event.DocumentEvent
import javax.swing.event.DocumentListener

class SettingsForm(private val initialState: LimitedWipSettings) {
    lateinit var root: JPanel

    private lateinit var watchdogPanel: JPanel
    private lateinit var watchdogEnabled: JCheckBox
    private lateinit var maxLinesInChange: JComboBox<*>
    private lateinit var notificationInterval: JComboBox<*>
    private lateinit var showRemainingInToolbar: JCheckBox
    private lateinit var noCommitsAboveThreshold: JCheckBox
    private lateinit var exclusions: RawCommandLineEditor

    private lateinit var autoRevertPanel: JPanel
    private lateinit var autoRevertEnabled: JCheckBox
    private lateinit var minutesTillRevert: JComboBox<*>
    private lateinit var notifyOnRevert: JCheckBox
    private lateinit var showTimerInToolbar: JCheckBox
    private lateinit var openReadme: LinkLabel<Void>

    private lateinit var tcrPanel: JPanel
    private lateinit var tcrEnabled: JCheckBox
    private lateinit var notifyOnTcrRevert: JCheckBox
    private lateinit var tcrActionOnPassedTest: JComboBox<*>

    private val currentState = LimitedWipSettings()
    private var isUpdatingUI: Boolean = false

    private val tcrActionByIndex = HashBiMap.create<Int, TcrAction>().also {
        it[0] = OpenCommitDialog
        it[1] = Commit
        it[2] = CommitAndPush
    }

    init {
        watchdogPanel.border = IdeBorderFactory.createTitledBorder("Change size watchdog")
        autoRevertPanel.border = IdeBorderFactory.createTitledBorder("Auto-revert")
        tcrPanel.border = IdeBorderFactory.createTitledBorder("TCR mode (test \\&\\& commit || revert)")
        exclusions = RawCommandLineEditor(ParametersListUtil.COLON_LINE_PARSER, ParametersListUtil.COLON_LINE_JOINER)
        exclusions.dialogCaption = "Resource patterns"

        currentState.loadState(initialState)
        updateUIFromState()

        fun doUpdate() {
            updateStateFromUI()
            updateUIFromState()
        }

        val commonActionListener = { _: ActionEvent -> doUpdate() }

        watchdogEnabled.addActionListener(commonActionListener)
        maxLinesInChange.addActionListener(commonActionListener)
        notificationInterval.addActionListener(commonActionListener)
        showRemainingInToolbar.addActionListener(commonActionListener)
        noCommitsAboveThreshold.addActionListener(commonActionListener)
        exclusions.document.addDocumentListener(object : DocumentListener {
            override fun changedUpdate(e: DocumentEvent?) = doUpdate()
            override fun insertUpdate(e: DocumentEvent?) = doUpdate()
            override fun removeUpdate(e: DocumentEvent?) = doUpdate()
        })

        autoRevertEnabled.addActionListener(commonActionListener)
        minutesTillRevert.addActionListener(commonActionListener)
        notifyOnRevert.addActionListener(commonActionListener)
        showTimerInToolbar.addActionListener(commonActionListener)

        tcrEnabled.addActionListener(commonActionListener)
        notifyOnTcrRevert.addActionListener(commonActionListener)
        tcrActionOnPassedTest.addActionListener(commonActionListener)

        openReadme.setListener(
            { _, _ -> BrowserUtil.open("https://github.com/dkandalov/limited-wip/blob/master/README.md#limited-wip") },
            null
        )
    }

    fun updateUIFromState() {
        if (isUpdatingUI) return
        isUpdatingUI = true

        watchdogEnabled.isSelected = currentState.watchdogEnabled
        maxLinesInChange.selectedItem = currentState.maxLinesInChange.toString()
        notificationInterval.selectedItem = currentState.notificationIntervalInMinutes.toString()
        showRemainingInToolbar.isSelected = currentState.showRemainingChangesInToolbar
        noCommitsAboveThreshold.isSelected = currentState.noCommitsAboveThreshold
        exclusions.text = currentState.exclusions

        autoRevertEnabled.isSelected = currentState.autoRevertEnabled
        minutesTillRevert.selectedItem = currentState.minutesTillRevert.toString()
        notifyOnRevert.isSelected = currentState.notifyOnRevert
        showTimerInToolbar.isSelected = currentState.showTimerInToolbar

        tcrEnabled.isSelected = currentState.tcrEnabled
        notifyOnTcrRevert.isSelected = currentState.notifyOnTcrRevert
        tcrActionOnPassedTest.selectedIndex = tcrActionByIndex.inverse()[currentState.tcrActionOnPassedTest]!!

        currentState.autoRevertEnabled.let {
            minutesTillRevert.isEnabled = it
            notifyOnRevert.isEnabled = it
            showTimerInToolbar.isEnabled = it
        }
        currentState.watchdogEnabled.let {
            maxLinesInChange.isEnabled = it
            notificationInterval.isEnabled = it
            showRemainingInToolbar.isEnabled = it
            noCommitsAboveThreshold.isEnabled = it
            exclusions.isEnabled = it
        }
        currentState.tcrEnabled.let {
            notifyOnTcrRevert.isEnabled = it
            tcrActionOnPassedTest.isEnabled = it
        }

        isUpdatingUI = false
    }

    private fun updateStateFromUI() {
        try {
            currentState.watchdogEnabled = watchdogEnabled.isSelected
            val lineCount = (maxLinesInChange.selectedItem as String).toInt()
            if (isValidChangedSizeRange(lineCount)) {
                currentState.maxLinesInChange = lineCount
            }
            var minutes = (notificationInterval.selectedItem as String).toInt()
            if (isValidNotificationInterval(minutes)) {
                currentState.notificationIntervalInMinutes = minutes
            }
            currentState.showRemainingChangesInToolbar = showRemainingInToolbar.isSelected
            currentState.noCommitsAboveThreshold = noCommitsAboveThreshold.isSelected

            currentState.exclusions = exclusions.text

            currentState.autoRevertEnabled = autoRevertEnabled.isSelected
            minutes = (minutesTillRevert.selectedItem as String).toInt()
            if (isValidMinutesTillRevert(minutes)) {
                currentState.minutesTillRevert = minutes
            }
            currentState.notifyOnRevert = notifyOnRevert.isSelected
            currentState.showTimerInToolbar = showTimerInToolbar.isSelected

            currentState.tcrEnabled = tcrEnabled.isSelected
            currentState.notifyOnTcrRevert = notifyOnTcrRevert.isSelected
            currentState.tcrActionOnPassedTest = tcrActionByIndex[tcrActionOnPassedTest.selectedIndex]!!

        } catch (ignored: NumberFormatException) {
        }
    }

    fun applyChanges(): LimitedWipSettings {
        initialState.loadState(currentState)
        return initialState
    }

    fun resetChanges() = currentState.loadState(initialState)

    fun isModified() = currentState != initialState
}
